src/core/pointcloud/qgspointcloudlayerrenderer.cpp

```diff
@@ -197,39 +198,61 @@ bool QgsPointCloudLayerRenderer::render()
   {
     canceled = !renderIndex( pc );
   }
-  else
+  else if ( const QgsVirtualPointCloudProvider *vpcProvider = dynamic_cast<QgsVirtualPointCloudProvider *>( mLayer->dataProvider() ) )
   {
-    mSubIndexExtentRenderer->startRender( context );
+    QVector< QgsPointCloudSubIndex > visibleIndexes;
     for ( const auto &si : mSubIndexes )
     {
-      if ( canceled )
-        break;
-
-      QgsPointCloudIndex *pc = si.index();
-
-      if ( !renderExtent.intersects( si.extent() ) )
-        continue;
-
-      if ( !pc || !pc->isValid() || renderExtent.width() > si.extent().width() )
+      if ( renderExtent.intersects( si.extent() ) )
       {
-        // when dealing with virtual point clouds, we want to render the individual extents when zoomed out
-        // and only use the selected renderer when zoomed in
-        mSubIndexExtentRenderer->renderExtent( si.polygonBounds(), context );
-        // render the label of point cloud tile
-        if ( mSubIndexExtentRenderer->showLabels() )
-        {
-          mSubIndexExtentRenderer->renderLabel(
-            context.renderContext().mapToPixel().transformBounds( si.extent().toRectF() ),
-            si.uri().section( "/", -1 ).section( ".", 0, 0 ),
-            context );
-        }
+        visibleIndexes.append( si );
+      }
+    }
+    const bool zoomedOut = renderExtent.width() > vpcProvider->averageSubIndexWidth() ||
+                           renderExtent.height() > vpcProvider->averageSubIndexHeight();
+    // if the overview of virtual point cloud exists, and we are zoomed out, we render just overview
+    if ( vpcProvider->overview() && zoomedOut &&
+         mRenderer->zoomOutBehavior() == Qgis::PointCloudZoomOutRenderBehavior::RenderOverview )
+    {
+      renderIndex( vpcProvider->overview() );
+    }
+    else
+    {
+      // if the overview of virtual point cloud exists, and we are zoomed out, but we want both overview and extents,
+      // we render overview
+      if ( vpcProvider->overview() && zoomedOut &&
+           mRenderer->zoomOutBehavior() == Qgis::PointCloudZoomOutRenderBehavior::RenderOverviewAndExtents )
+      {
+        renderIndex( vpcProvider->overview() );
       }
-      else
+      mSubIndexExtentRenderer->startRender( context );
+      for ( const auto &si : visibleIndexes )
```

Comment:
QGIS coding convention is to avoid `auto` for types like this -- please use an explicit type instead

Reply:
I actually reused what was already there, but I'll refactor the code as this pops up in multiple places.